cpp/include/cudf/join.hpp

```diff
@@ -438,6 +446,64 @@ class hash_join {
   const std::unique_ptr<impl_type const> _impl;
 };
 
+/**
+ * @brief Unique hash join that builds hash table in creation and probes results in subsequent
+ * `*_join` member functions
+ *
+ * @note Behavior is undefined if the build table contains duplicates.
+ * @note All NaNs are considered as equal
+ *
+ * @tparam HasNested Flag indicating whether there are nested columns in build/probe table
+ */
+// TODO: `HasNested` to be removed via dispatching
+template <has_nested HasNested = has_nested::NO>
+class unique_hash_join {
+ public:
+  unique_hash_join() = delete;
+  ~unique_hash_join();
+  unique_hash_join(unique_hash_join const&)            = delete;
+  unique_hash_join(unique_hash_join&&)                 = delete;
+  unique_hash_join& operator=(unique_hash_join const&) = delete;
+  unique_hash_join& operator=(unique_hash_join&&)      = delete;
+
+  /**
+   * @brief Constructs a unique hash join object for subsequent probe calls
+   *
+   * @param build The build table that contains unique elements
+   * @param probe The probe table, from which the keys are probed
+   * @param has_nulls Flag to indicate if there exists any nulls in the `build` table or
+   *        any `probe` table that will be used later for join
+   * @param compare_nulls Controls whether null join-key values should match or not
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   */
+  unique_hash_join(cudf::table_view const& build,
+                   cudf::table_view const& probe,
```

Comment:
question: Why is the probe table part of the constructor? That shouldn't be needed for creating the initial hash table?

Reply:
I wondered the same thing, seems to be a temporary limitation in cuco.  See discussion at https://github.com/rapidsai/cudf/pull/14990#discussion_r1482039020

Reply:
@jrhemstad This is the issue I brought up yesterday with you and Daniel: we have to have the probe table info when constructing the hash table since the new `cuco::static_set` takes the two-table comparator type as the `KeyEqual` template parameter.

Reply:
In that case, there isn't much value in providing an object-oriented interface for `unique_hash_join` yet, right? The purpose of the OO interface was to allow building a hash table once and probe it multiple times with other tables.

I'd say just provide a free function version of this functionality until the equality thing is sorted out.

Reply:
I think we agreed that customizing equality for a non-modifying operation is acceptable and adding `static_set::retrieve` taking a custom key equal is fairly simple on the cuco side (since the majority of the work is to move the kernel and device functions in this PR to cuco) and the feature will be ready in no more than two weeks.

I'm fine with the free function solution but making it a free function for now and then reverting it back to the current OO interface in two weeks adds quite a bit back-and-forth efforts.